### PR TITLE
Filter wordless appeals alongside too-short ones

### PR DESCRIPTION
**Pull request by holdenk**
## Summary
Extends appeal validation to reject not just short appeals, but also appeals that are long enough but contain no actual words—e.g., a model echoing back claim numbers, dates, and dollar amounts. This prevents junk output from bypassing the minimum-length check.

## Key Changes

- **New utility functions in `utils.py`:**
  - `appeal_word_count()` — counts word-like tokens (runs of 2+ letters) in text
  - `appeal_has_words()` — checks if text contains enough words to be a real letter
  - `describe_unusable_appeal()` — explains why text fails validation (too-short vs. not-words)
  - Renamed `warn_too_short_appeal()` → `warn_unusable_appeal()` to cover both rejection reasons

- **Updated `is_real_appeal()` validation:**
  - Now requires both minimum length AND actual words (via `appeal_has_words()`)
  - Handles unspaced scripts (Chinese, Japanese, Thai) by accepting one long run of letters as prose

- **Consistent filtering across all appeal sources:**
  - Streaming generation path (`common_view_logic.py`)
  - Synthesis output (`common_view_logic.py`)
  - Fallback detection in `generate_appeal.py`
  - Previously-saved appeals in the database

- **Improved logging:**
  - Single warning format for all rejection reasons
  - Distinguishes "too-short" from "not-words" in output
  - Reports word count alongside character count for wordless rejects

## Implementation Details

- Word detection uses a Unicode-aware regex (`[^\W\d_]{2,}`) that excludes digits, punctuation, and lone letters
- Minimum word threshold is deliberately low (2 words) to catch worst-case junk while allowing legitimate short appeals
- Non-Latin scripts are supported: a single unspaced run of 15+ letters (e.g., Chinese text) counts as prose
- All appeal validation now goes through `is_real_appeal()`, ensuring consistent behavior across generation, synthesis, and database retrieval

## Tests Added
- `test_synthesis_wordless_output_is_filtered()` — synthesized junk is rejected
- `test_existing_wordless_appeal_is_skipped()` — saved junk is not re-delivered
- `test_wordless_first_returns_none()` — fallback triggers on wordless first item
- Parametrized tests for `appeal_has_words()` and `appeal_word_count()` with edge cases
- Refactored test helper `_run_generate_appeals_over_saved()` to reduce duplication

https://claude.ai/code/session_01Xrb3ZyBDnQcWrDbC9pvvXj

## Summary by CodeRabbit

* **Bug Fixes**
  * Strengthened appeal validation to reject outputs that are too short or have no meaningful word-like content (including punctuation- and number-only text).
  * Prevented undeliverable/invalid generated and previously saved appeals from being returned.
  * Improved generation fallback when the first proposed appeal is unusable.
  * Enhanced reconnect handling so speculative/preview items flush and are served without duplicates when applicable.
* **Diagnostics**
  * Updated warnings to consistently distinguish “too-short” vs “not-words” unusable appeals with clearer context.

### CODE REVIEW

File: fighthealthinsurance/utils.py [hunk 1]
```diff
@@ -54,6 +54,20 @@
...
+# Deliberately a floor for the worst-of-the-worst (a model echoing back the
+# denial's identifiers instead of writing), NOT a quality bar: real appeals
+# run to a couple hundred words, so anything with fewer than two words is
+# junk by any measure.
+MIN_APPEAL_WORDS = 2
+
+# A word-like token: a run of two or more letters. Digits, punctuation, and
+# lone letters are excluded, so "99213", "$1,250.00", "11/02/2026" and the
+# "A" in "Plan A" contribute nothing.
+_APPEAL_WORD_RE = re.compile(r"[^\W\d_]{2,}")
```
[the remark this answers is not in the document]
Comment (holdenk) [reply fighthealthinsurance/utils.py]: Good catch — fixed in dfdd41a.

A word-like token now has to not be welded to a digit or underscore on either side:

```python
_APPEAL_WORD_RE = re.compile(r"(?<!\w)[^\W\d_]{2,}(?!\w)")
```

`AB123456789 CD987654321` and `H5521-001 BCBS12345 99213 $1,250.00` now score 0 words and are rejected.

Ordinary prose is unaffected — hyphenated words still split into two tokens (`well-documented`), possessives count once (`patient's`), and `COVID-19` still counts. I re-measured the six static appeal templates in `generate_appeal.py`: they drop from 276/196/206/197/225/317 tokens to 272/192/202/193/221/313, so ~4 tokens out of ~200 each, nowhere near the floor of 2.

Added regression cases for both the identifier forms and the hyphenated/possessive prose that must keep counting.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/async-unit/test_appeals_backend_citations.py [hunk 2]
```diff
@@ -174,6 +174,75 @@ def route(fn, *args, **kwargs):
...
+            "fighthealthinsurance.common_view_logic.interleave_iterator_for_keep_alive",
+            side_effect=passthrough_interleave,
+        ),
+        patch(
+            "fighthealthinsurance.common_view_logic.ProposedAppeal",
+        ) as mock_pa_cls,
+        patch(
+            "fighthealthinsurance.common_view_logic.appealGenerator"
+        ) as mock_appeal_gen,
+    ):
```
[the remark this answers is not in the document]
Comment (holdenk) [reply tests/async-unit/test_appeals_backend_citations.py:260]: Valid — fixed in 6246a6f.

`generate_appeals` awaits `get_rag_context_for_denial` under `asyncio.wait_for(..., timeout=30)` (`common_view_logic.py:3207`), and `RAGClient` defaults to `http://localhost:8001`, so the tests were relying on connection-refused to move on.

Applied the suggested patch target (`common_view_logic.get_rag_context_for_denial` — the module-level import, not a method on `AppealsBackendHelper` as the AI-agent prompt suggests). I applied it to all three harnesses in the file rather than only the one this PR added: `test_synthesis_threshold` and `test_synthesis_skips_verbatim_duplicate` had the identical gap, and fixing one while leaving two live would be arbitrary.

async-unit suite: 1399 passed.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: fighthealthinsurance/utils.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (holdenk) [reply fighthealthinsurance/utils.py]: Follow-up: the latest review round found three more real issues, all fixed in a1d1a56. Summarizing here since two of them were posted outside the diff range and have no thread of their own.

**Combining marks (confirmed, Major).** `[^\W\d_]` doesn't match Unicode category M, so Thai/Indic vowel and tone signs split words into single-letter runs — `appeal_has_words("ที่นี่ดี ที่นี่ดี ที่นี่ดี")` was `False` despite 24 meaningful characters. That's the worst failure mode for this filter: dropping a real appeal. Tokenizing now strips marks first, leaving base letters (`ที่นี่ดี` → `ทนด`). Verified English/ASCII tokenizes identically either way, and added Thai + Devanagari cases.

**Atomic promotion (confirmed, Major).** Fair catch that the reconnect change makes the overlap likely — the dropped socket's generator can still be draining while the replacement flushes the reserve, and `served_texts` is per-run so it can't dedupe across flows. Both promotion sites (the early flush and the end-of-flow reconciliation) now claim with a conditional `aupdate(speculative=False)` filtered on `speculative=True, chosen=False`, and the loser skips the row. Test added that drives the losing side of the race.

**`doQuery` payload type (confirmed).** `appeals.html:149` passes an object literal, so the `Map<string, string>` annotation was wrong. Worth noting the risk was pre-existing rather than introduced by the spread: `JSON.stringify(new Map())` is `"{}"`, so a real Map would have sent an empty payload long before this PR. Retyped as `Record<string, string>`; frontend rebuilt.

Also took the two nitpicks: synthesis-input query routed through `deliverable_candidates`, and the duplicated loguru sink setup lifted onto the existing `capture_logs` helper.

**Skipped:** the `tests/async-unit/test_dev_storage_settings.py` finding — this branch doesn't touch that file; it entered the diff range via the rebase onto main.

Suites: sync 1700, async 472, async-unit 1424, sync-actor 17, plus mypy, black, and `npm run build`.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: fighthealthinsurance/utils.py [hunk 3]
```diff
@@ -74,14 +93,61 @@ def meaningful_appeal_length(text: Optional[str]) -> int:
...
+    vowel and tone signs, for instance -- so leaving them in would break
+    "ที่นี่ดี" into single-letter runs and score a perfectly ordinary appeal as
+    wordless. Removing them leaves the base letters, which is all a token count
+    needs. Text without marks (including English) tokenizes identically either
+    way.
+    """
+    if not isinstance(text, str):
+        return []
+    unmarked = "".join(ch for ch in text if unicodedata.category(ch)[0] != "M")
+    return _APPEAL_WORD_RE.findall(unmarked)
```
[the remark this answers is not in the document]
Comment (holdenk) [reply fighthealthinsurance/utils.py]: Confirmed and fixed in 6ad065a — and it was worse than the description suggests. I measured before fixing:

```
"ที่นี่ดีที่นี่ดีที่นี่ดี"  raw_meaningful=24  stripped token='ทนดทนด...' len=9
```

So a 24-character unspaced Thai clause — comfortably over `MIN_APPEAL_CHARS` — was failing a 15-character bar on its 9 base letters. My mark-stripping fixed token *counting* and then shortchanged the length check for exactly the scripts it was added for.

Tokenizing now returns `(token, width)` pairs, where `width` counts each dropped mark back onto the letter it attached to, and the unspaced fallback tests `width`. Text without marks is unaffected — every width is then the token's own length.

I did consider the shorter fix of counting letters+marks across the whole text, but rejected it: `"1234567890 ab 12345"` would then pass on the strength of one two-letter word (16 letter-ish characters total). The per-token width keeps that at `False`. Verified across the full matrix — spaced/unspaced Thai, Devanagari, CJK, Latin, identifier echo, numbers-only, and the one-short-word-amid-digits case — plus regression cases for the 24-char unspaced clause and an 8-character one that must still be rejected.

The ast-grep XPath-injection warning on `_APPEAL_WORD_RE.findall(...)` is a false positive; that's a compiled `re` pattern, not an XPath expression.

Suites: sync 1700, async 472, async-unit 1426, plus mypy and black.

---
_Generated by [Claude Code](https://claude.ai/code)_
